src/persist-client/src/cache.rs

```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
+                    match res.message {
+                        Some(proto_pub_sub_message::Message::PushDiff(diff)) => {
+                            cache.metrics.pubsub_client.receiver.push_received.inc();
+                            let shard_id = diff.shard_id.into_rust().expect("WIP");
+                            let diff = VersionedData {
+                                seqno: diff.seqno.into_rust().expect("WIP"),
+                                data: diff.diff,
+                            };
+                            debug!(
+                                "client got diff {} {} {}",
+                                shard_id,
+                                diff.seqno,
+                                diff.data.len()
+                            );
+                            cache.apply_diff(&shard_id, diff);
+                            let now = SystemTime::now()
+                                .duration_since(SystemTime::UNIX_EPOCH)
+                                .expect("failed to get millis since epoch")
+                                .as_micros();
+                            let latency = now.saturating_sub(timestamp) as f64;
+                            cache
+                                .metrics
+                                .pubsub_client
+                                .receiver
+                                .approx_diff_latency
+                                .observe(latency);
+                        }
+                        ref msg @ None | ref msg @ Some(_) => {
+                            warn!("pubsub client received unexpected message: {:?}", msg);
+                            cache
+                                .metrics
+                                .pubsub_client
+                                .receiver
+                                .unknown_message_received
+                                .inc();
+                        }
+                    }
+                }
+                cache.metrics.pubsub_client.receiver.connected.set(0);
+            });
+        } else {
+            state_cache.metrics.pubsub_client.receiver.connected.set(0);
         }
+
+        state_cache
     }
 
     #[cfg(test)]
-    pub(crate) fn new_no_metrics() -> Self {
-        Self::new(Arc::new(Metrics::new(
+    pub(crate) fn new_no_metrics() -> Arc<Self> {
+        Self::new(
             &PersistConfig::new_for_tests(),
-            &MetricsRegistry::new(),
-        )))
+            Arc::new(Metrics::new(
+                &PersistConfig::new_for_tests(),
+                &MetricsRegistry::new(),
+            )),
+            None,
+        )
+    }
+
+    pub(crate) fn apply_diff(&self, shard_id: &ShardId, diff: VersionedData) {
+        if let Some(state) = self.get_cached(shard_id) {
+            // WIP: could register callbacks here. if not applied, fallback
+            let applied = state.push_diff(diff);
+            let shard_metrics = self.metrics.shards.shard(shard_id);
```

Comment:
feels somewhat unfortunate to have to look up the metrics by shard_id, too, because my mind immediately jumps to thinking about race conditions. seems not-crazy to put the ShardMetrics on LockingTypedState and make it available via DynState?

Reply:
done!